madmom/ml/hmm.pyx

```diff
@@ -427,6 +428,113 @@ class HiddenMarkovModel(object):
+            for state in range(num_states):
+                # reset the current viterbi variable
+                current_viterbi[state] = -INFINITY
+                # get the observation model probability density value
+                # the om_pointers array holds pointers to the correct
+                # observation probability density value for the actual state
+                # (i.e. column in the om_densities array)
+                # Note: defining density here gives a 5% speed-up!?
+                density = om_densities[frame, om_pointers[state]]
+                # iterate over all possible previous states
+                # the tm_pointers array holds pointers to the states which are
+                # stored in the tm_states array
+                for pointer in range(tm_pointers[state],
+                                     tm_pointers[state + 1]):
+                    # get the previous state
+                    prev_state = tm_states[pointer]
+                    # weight the previous state with the transition probability
+                    # and the current observation probability density
+                    transition_prob = previous_viterbi[prev_state] + \
+                                      tm_probabilities[pointer] + density
+                    # if this transition probability is greater than the
+                    # current one, overwrite it and save the previous state
+                    # in the back tracking pointers
+                    if transition_prob > current_viterbi[state]:
+                        # update the transition probability
+                        current_viterbi[state] = transition_prob
+                        # update the back tracking pointers
+                        bt_pointers[frame, state] = prev_state
+
+            # overwrite the old states with the current ones
+            cur_id += 1
+            previous_viterbi = viterbi_vars[cur_id - 1]
+            current_viterbi = viterbi_vars[cur_id]
+
+        # fetch the final best state
+        state = np.asarray(current_viterbi).argmax()
+        # set the path's probability to that of the best state
+        log_probability = current_viterbi[state]
+        # back tracked path, a.k.a. path sequence
+        path = np.empty(num_observations, dtype=np.uint32)
+        # track the path backwards, start with the last frame and do not
+        # include the pointer for frame 0, since it includes the transitions
+        # to the prior distribution states
+        for frame in range(num_observations -1, -1, -1):
+            # save the state in the path
+            path[frame] = state
+            # fetch the next previous one
+            state = bt_pointers[frame, state]
+
+        # return the tracked path and its probability
+        return path, log_probability, np.asarray(bt_pointers), np.asarray(viterbi_vars)
+
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
```

Comment:
Is adding this `verbose_viterbi()` a mistake or on purpose? If on purpose, I'd propose to add a verbose parameter to `viterbi()` instead.